File: ios/DutchLearnerPackage/Sources/DutchLearnerPackage/Utilities/FuzzyMatch.swift
//
//  FuzzyMatch.swift
//  DutchLearner
//
//  Fuzzy string matching with Levenshtein distance for answer validation
//

import Foundation

public enum FuzzyMatch {
    /// Calculate Levenshtein distance between two strings
    private static func levenshteinDistance(_ str1: String, _ str2: String) -> Int {
        let s1 = Array(str1)
        let s2 = Array(str2)

        var matrix = Array(repeating: Array(repeating: 0, count: s1.count + 1), count: s2.count + 1)

        // Initialize first column and row
        for i in 0...s1.count {
            matrix[0][i] = i
        }

        for j in 0...s2.count {
            matrix[j][0] = j
        }

        // Fill matrix
        for j in 1...s2.count {
            for i in 1...s1.count {
                let indicator = s1[i - 1] == s2[j - 1] ? 0 : 1
                matrix[j][i] = min(
                    matrix[j][i - 1] + 1, // deletion
                    matrix[j - 1][i] + 1, // insertion
                    matrix[j - 1][i - 1] + indicator // substitution
                )
            }
        }

        return matrix[s2.count][s1.count]
    }

    /// Normalize string for comparison
    private static func normalize(_ str: String) -> String {
        return str
            .lowercased()
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .replacingOccurrences(of: "[^\\w\\s]", with: "", options: .regularExpression) // Remove punctuation
            .replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression) // Normalize whitespace
    }

    /// Check if two strings are similar enough
    /// - Parameters:
    ///   - userInput: User's input string
    ///   - correctAnswer: Expected correct answer
    ///   - threshold: Maximum number of character differences allowed (default 2)
    /// - Returns: True if strings match within threshold
    public static func fuzzyMatch(userInput: String, correctAnswer: String, threshold: Int = 2) -> Bool {
        guard !userInput.isEmpty, !correctAnswer.isEmpty else {
            return false
        }

        let normalizedInput = normalize(userInput)
        let normalizedAnswer = normalize(correctAnswer)

        // Exact match after normalization
        if normalizedInput == normalizedAnswer {
            return true
        }

        // Handle multiple correct answers separated by '/' or ','
        let separatorPattern = "[/,]"
        let possibleAnswers = normalizedAnswer
            .components(separatedBy: CharacterSet(charactersIn: "/,"))
            .map { $0.trimmingCharacters(in: .whitespaces) }

        for answer in possibleAnswers {
            if normalizedInput == answer {
                return true
            }

            // Check Levenshtein distance
            let distance = levenshteinDistance(normalizedInput, answer)
            let maxLength = max(normalizedInput.count, answer.count)

            // Allow up to threshold character differences, but not more than 30% of the word length
            if distance <= threshold && Double(distance) <= Double(maxLength) * 0.3 {
                return true
            }
        }

        return false
    }

    /// Get similarity score between 0 and 1
    /// - Parameters:
    ///   - str1: First string
    ///   - str2: Second string
    /// - Returns: Similarity score (1.0 = identical, 0.0 = completely different)
    public static func similarityScore(_ str1: String, _ str2: String) -> Double {
        let normalized1 = normalize(str1)
        let normalized2 = normalize(str2)

        let maxLength = max(normalized1.count, normalized2.count)
        if maxLength == 0 { return 1.0 }

        let distance = levenshteinDistance(normalized1, normalized2)
        return 1.0 - Double(distance) / Double(maxLength)
    }
}

File: src/utils/fuzzyMatch.ts
// Calculate Levenshtein distance between two strings
const levenshteinDistance = (str1: string, str2: string): number => {
  const matrix = Array(str2.length + 1)
    .fill(null)
    .map(() => Array(str1.length + 1).fill(null));

  for (let i = 0; i <= str1.length; i++) {
    matrix[0][i] = i;
  }

  for (let j = 0; j <= str2.length; j++) {
    matrix[j][0] = j;
  }

  for (let j = 1; j <= str2.length; j++) {
    for (let i = 1; i <= str1.length; i++) {
      const indicator = str1[i - 1] === str2[j - 1] ? 0 : 1;
      matrix[j][i] = Math.min(
        matrix[j][i - 1] + 1, // deletion
        matrix[j - 1][i] + 1, // insertion
        matrix[j - 1][i - 1] + indicator, // substitution
      );
    }
  }

  return matrix[str2.length][str1.length];
};

// Normalize string for comparison
const normalizeString = (str: string): string => {
  return str
    .toLowerCase()
    .trim()
    .replace(/[^\w\s]/g, "") // Remove punctuation
    .replace(/\s+/g, " "); // Normalize whitespace
};

// Check if two strings are similar enough
export const fuzzyMatch = (
  userInput: string,
  correctAnswer: string,
  threshold: number = 2,
): boolean => {
  if (!userInput || !correctAnswer) {
    return false;
  }

  const normalizedInput = normalizeString(userInput);
  const normalizedAnswer = normalizeString(correctAnswer);

  // Exact match after normalization
  if (normalizedInput === normalizedAnswer) {
    return true;
  }

  // Handle multiple correct answers separated by '/' or ','
  const possibleAnswers = normalizedAnswer
    .split(/[\/,]/)
    .map((answer) => answer.trim());

  for (const answer of possibleAnswers) {
    if (normalizedInput === answer) {
      return true;
    }

    // Check Levenshtein distance
    const distance = levenshteinDistance(normalizedInput, answer);
    const maxLength = Math.max(normalizedInput.length, answer.length);

    // Allow up to 2 character differences, but not more than 30% of the word length
    if (distance <= threshold && distance <= maxLength * 0.3) {
      return true;
    }
  }

  return false;
};

// Get similarity score between 0 and 1
export const getSimilarityScore = (str1: string, str2: string): number => {
  const normalized1 = normalizeString(str1);
  const normalized2 = normalizeString(str2);

  const maxLength = Math.max(normalized1.length, normalized2.length);
  if (maxLength === 0) return 1;

  const distance = levenshteinDistance(normalized1, normalized2);
  return 1 - distance / maxLength;
};
